### DISCUSSION TITLE
2023-08-21 정기 백엔드 개발 회의

### DISCUSSION THREAD

**Discussion by scarf005** [jiphyeonjeon-42/backend#677] [General]
안건을 자유롭게 추가해주세요.

**Comment by scarf005** [comment]
## 배포 일정에 대해

버그픽스가 포함돼 있어 #675 를 빨리 머지하면 좋을 것 같습니다.

**Reply by jimin52** [reply]
지금 배포가 안 되고 있는데 codeQL 친구 무시하고 머지할 방법이 없을까요?

**Reply by scarf005** [reply]
![image](https://github.com/jiphyeonjeon-42/backend/assets/54838975/dd1aa06e-4930-4011-aa35-debbcb6e5524)

codeQL리뷰는 생략 가능하나 리뷰 approve가 없어 머지가 되지 않는 것으로 보입니다.

**Reply by scarf005** [reply]
완료했습니다!

**Comment by jimin52** [comment]
## rate-limit 미들웨어를 추가하는 이슈 생성
code scanner 가 경고한 dos 공격에 대비한 rate-limit 미들웨어의 도입 필요한 것 같습니다.

[참고 자료](https://www.npmjs.com/package/express-rate-limit)
[해당 이슈](https://github.com/jiphyeonjeon-42/backend/security/code-scanning/6)

**Reply by scarf005** [reply]
개발 서버에서 하는 방법처럼 nginx에 rate limit을 거는 방식도 괜찮을 것 같습니다.

https://github.com/jiphyeonjeon-42/backend/blob/6f1957bf271103ac00f112d7561a77db8397fce0/nginx.dev/nginx.conf#L28

https://github.com/jiphyeonjeon-42/backend/blob/6f1957bf271103ac00f112d7561a77db8397fce0/nginx.dev/conf.d/default.conf#L40

참고 링크
- https://www.nginx.com/blog/rate-limiting-nginx/
- https://willseungh0.tistory.com/191

**Reply by jimin52** [reply]
책 조회의 경우 타자를 칠때마다 api 콜이 날아오는 것으로 알고 있습니다.
CRUD 중 Read 와 CUD 의 rate limit 을 따로 걸던가
하나로 통합한다고 해도 모든 rate limit 을 넉넉하게 걸어놓는 것이 좋을 것 같습니다.
기본값으로 초당 10회정도가 좋을 것 같아요

**Reply by scarf005** [reply]
네, nginx 라우트에서 경로별로 rate limit해주면 될 것 같습니다.

별개로, 검색창 기능을 프론트엔드에서 처리하는 방식도 DB 쿼리 수 감소 측면에서 고려해 볼 만해 보입니다. 프론트엔드에서 매일 [indexedDB](https://developer.mozilla.org/ko/docs/Web/API/IndexedDB_API)에 전체 도서 목록을 캐싱해 두었다가, 매 타자마다 캐싱한 목록에서 [fuzzy search](https://github.com/krisk/Fuse)를 하는 방식으로요. 다만 아직은 백엔드에서 검색을 처리하는 것이 더 간단할 것 같네요.
